### ISSUE TITLE
Disallow $ in non-tuple syntax

### ISSUE THREAD

**Issue by gzm0** [scala-records/scala-records#121]
See #118 and #110

**Comment by cvogt** [comment]
is that only for applyDynamicNamed-construction? Still allowed in member access I suppose? `someRecord.$eq`

I'd also like to support the blackbox construction syntax:
```scala
Record(new{def name = "Chris"; def $inPocket = 35})
```
I think here we should support it. Also the encoding problems of applyDynamicNamed shouldn't be an issue here.

**Comment by gzm0** [comment]
Hmmm... it should probably also be disallowed in member access... Otherwise `rec.$eq` will actually query the `=` field, not the `$eq` field...

If you want blackbox construction syntax, there it should be ok, yes.

**Comment by cvogt** [comment]
mh... I think uh. `$eq` shouldn't query the `=` field, but the `$eq` field. Also it seems reasonable to assume to me, that encode/decode should be each others inverses for strings not containing `$`. In that case applyDynamicNamed should immediately decode and everything else should keep names as they are. Does that work for ```r.`-````?

**Comment by gzm0** [comment]
Oh... yes, it does. Actually the problem really is on Scala's side. I though the dynamic things always give you the encoded identifiers, but that's actually not true.

So yes `$` should be allowed in member access.

**Comment by gzm0** [comment]
I'm having the feeling that we are unable to disallow it, since we wont know where it is coming from... We'll see...

**Comment by cvogt** [comment]
in constructors? the applyDynamicNamed impl can just check, no?

**Comment by gzm0** [comment]
No. Because it will see: "$eq"... was it "=" which was translated or was "$eq" written there?

**Comment by gzm0** [comment]
FAIL

**Comment by cvogt** [comment]
argh..

**Comment by cvogt** [comment]
does this pattern have a name? CLBI? Cross-Library-Bug-Injection?

**Comment by gzm0** [comment]
It's called Scala

**Comment by cvogt** [comment]
**S**uper **c**ross **a**pplication **l**ibrary **a**ttack :(

**Comment by cvogt** [comment]
alright, if we can't prevent the use of `$` we should document it well, expect it not to be used and ideally open a ticket and ideally PR to fix it in scalac. I would actually classify the current applyDynamicNamed behavior as a bug.

If I am not mistaken the responsible snippet in the compiler is
https://github.com/scala/scala/blob/fe05bc9951f9e89bac59230932e50ec850df6ce4/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3998

@vjovanov @gzm0 @clhodapp @lrytz anyone aware of an existing ticket or PR? couldn't find one. Otherwise I'd open one.

Repeating the issue
```
scala> import scala.language.dynamics
scala> object Foo extends Dynamic{ def applyDynamicNamed(m: String)(args:(String, Any)*) = println(args) }
scala> Foo(`foo-$minusbar`=5)
````
Expected result:
```
WrappedArray((foo-$minusbar,5))
```
Actual result:
```
WrappedArray((foo$minus$minus,5))
```
meaning libraries like scala-records can't distinguish `-` from `$minus`, can't undo it nor prevent the use of `$`. Hence it's a blocker for #121
